### ISSUE TITLE
Plotnine performance issue

### ISSUE THREAD

**Issue by bevankoopman** [has2k1/plotnine#132]
Firstly, thanks for creating this project - I've been looking for a python grammar of graphics implementation for a while.

I'm having a performance issue plotting a large Pandas data frame.

The data is found in `prob_w_d_df`, which is a `(600000, 3)` data frame; `prob_w_d_df.describe()` gives:

```
	doc_id		prob
	------		----
count	600000		6.000000e+05
mean	29		1.000000e-04
std	17.318117	4.862052e-04
min	0.000000	2.285988e-07
25%	14.750000	1.205222e-06
50%	29.500000	5.755733e-06
75%	44.250000	3.625872e-05
max	59.000000	9.012307e-02
```

The plot code is:

```python
ggplot(prob_w_d_df, aes(x='feature', y='prob', group=1)) + 
 geom_line() + 
 xlab("Word") + 
 ylab("P(word)") +
 facet_wrap('~doc_id', scales="free") +
 theme_minimal(base_family="Arial") +
 theme(figure_size=(15,15), axis_text_x=element_blank())
) 
```

This takes **1,543 seconds** (~27 minutes) to generate the plot.

Now I understand it's a large data frame so will be slow; however, when I replicate the exact above code with `ggplot2` in R it takes **65 seconds**. I understand R may be faster but the difference seems to great.

Is there anything I'm doing wrong or should do you improve performance?

(I've shared a CSV dump of `prob_w_d_df` if it helps: [`prob_w_d_df.csv`](https://drive.google.com/open?id=17rr13kH5uS2gwkf_-K9m60N0gWUwoy24).)

Thanks.

**Comment by has2k1** [comment]
That is a big difference and I don't have the resources to investigate it at the moment. Can you post the result, you can use one from R.

Also, can you time how long it takes to build the plot

```python
p = ggplot(prob_w_d_df, aes(x='feature', y='prob', group=1)) + 
 geom_line() + 
 xlab("Word") + 
 ylab("P(word)") +
 facet_wrap('~doc_id', scales="free") +
 theme_minimal(base_family="Arial") +
 theme(figure_size=(15,15), axis_text_x=element_blank())
)
p._build() 
```

I suspect that it is the drawing (MPL) part that is slow.

**Comment by bevankoopman** [comment]
Thanks for the quick reply.

Attached is the generated plot in R:
![rplot](https://user-images.githubusercontent.com/3433630/38963993-f6030f9e-43b7-11e8-9e0b-07f9be1ffb4c.png)

The timing for the above `_build()` call was, on average, only **15 seconds** (I ran a few iterations).

**Comment by has2k1** [comment]
I thought that it would have many panels; Matplotlib struggles to draw many axes, but I'll have to find out if that is responsible for all of the 27 minutes. Plus,  find out the type of relationship between the number of axes and the performance.

**Comment by has2k1** [comment]
**Notes**

- [Problematic lines](https://github.com/has2k1/plotnine/blob/7b129e25fafe1371ea3ac541323f028904f91fc7/plotnine/geoms/geom_path.py#L62-L65). 
- `DataFrame.apply` is/can be slow, especial using normal python functions like `all` (line 65).

**Comment by grgmiller** [comment]
I was having a similar issue with performance when plotting data from a Pandas dataframe. What I found was that plotnine was having trouble if one of my variables was an object datatype. My x-axis was a timeseries of minutes of the day, which I had converted to strings, like "12:05", "12:10", etc. 

I found that when I converted these timestamps to float datatypes, like  12.0833, 12.1667, etc, it improved the performance considerably.
